### PR TITLE
Better logs and error messages

### PR DESCRIPTION
**Pull request by cyrille-leclerc**
Better logs and error messages. Especially for SSL and invalid credentials.

### CODE REVIEW

**Comment by cyrille-leclerc** [comment]
@reviewbybees

**Comment by jglick** [comment]
:bee:

**Comment by recena** [comment]
@cyrille-leclerc Before to merge this one, I'm reviewing this https://github.com/jenkinsci/github-branch-source-plugin/pull/23 it is related.

**Comment by recena** [comment]
DON NOT MERGE!

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 1]
```diff
@@ -107,12 +109,12 @@ public String getApiUri() {
...
         GitHub github = Connector.connect(apiUri, credentials);
         if (credentials != null && !github.isCredentialValid()) {
-            listener.getLogger().format("Invalid scan credentials, skipping%n");
+            listener.getLogger().format("Invalid scan credentials %s to connect to %s, skipping%n", CredentialsNameProvider.name(credentials), apiUri == null ? "github.com" : apiUri);
             return;
         }
 
         if (!github.isAnonymous()) {
-            listener.getLogger().format("Connecting to GitHub using %s%n", CredentialsNameProvider.name(credentials));
+            listener.getLogger().format("Connecting to %s using %s%n", apiUri == null ? "github.com" : apiUri, CredentialsNameProvider.name(credentials));
```
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Use `HttpsRepositoryUriResolver.hostnameFromApiUri()`

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 2]
```diff
@@ -107,12 +109,12 @@ public String getApiUri() {
         StandardCredentials credentials = Connector.lookupScanCredentials(observer.getContext(), apiUri, scanCredentialsId);
         GitHub github = Connector.connect(apiUri, credentials);
         if (credentials != null && !github.isCredentialValid()) {
-            listener.getLogger().format("Invalid scan credentials, skipping%n");
+            listener.getLogger().format("Invalid scan credentials %s to connect to %s, skipping%n", CredentialsNameProvider.name(credentials), apiUri == null ? "github.com" : apiUri);
```
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Ditto

**Comment by recena** [comment]
:ant: because there is an explicit method to resolve the hostname from an apiUrl.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 1 above)
Comment (cyrille-leclerc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: @recena `RepositoryUriResolver#hostnameFromApiUri` just returns the hostname when I would like to display the entire API URL. I suspect that some people will make a mistake defining the API URL and will be happy to see the entire URL.

BTW I did not catch why `GitHubSCMNavigator#apiUri` is nullable with a kind of discovery happening latter on.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 2 above)
Comment (cyrille-leclerc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Ditto as well :-)

**Comment by cyrille-leclerc** [comment]
I would prefer to have the entire API URL including port, URI & protocol rather than just the hostname.

`404 invoking github.example.com` is less helpful than `404 invoking https://github.example.com/api/v3_I_introduced_a_typo`

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 1 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: @cyrille-leclerc This only makes sense when you are working with a GHE instance. When you configure an instance, its endpoint has always the same format: `http(s)://hostname/api/v3/`. For that reason, the important part is the `hostname`.

**Comment by recena** [comment]
@cyrille-leclerc It should not happen if the form validation for GitHub Enterprise instances is working fine.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java (hunk 2 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: If we reach out this point we need to see in the console if we are working with github.com or our owner GHE instance.

**Comment by recena** [comment]
@cyrille-leclerc I don't want to block this PR but my opinion is on the table.

**Comment by recena** [comment]
:bee:

**Comment by recena** [comment]
@reviewbybees done

**Comment by cyrille-leclerc** [comment]
> It should not happen

I wish you were with me when I tried to understand why things went wrong in my setup.

I propose that we lower the level of details of the logs once the code stops silently ignoring critical exception like `IOException` (using `catch (IOException) {// ignore because this cannot not happen}`) .

FYI even org.kohsuke:github-api silently swallows IOException. 

I'll show you how hard it is to findout that an SSL cert is not trusted.

Places where we silently swallow exceptions:

* https://github.com/jenkinsci/github-branch-source-plugin/blob/0765e53fa47aa3ba27ace80e414ca12c0bcad4a7/src/main/java/org/jenkinsci/plugins/github_branch_source/RepositoryUriResolver.java#L42
   * `// ignore`: ignore a MalformedURLException? I would prefer to say that it's an IllegalArgument or and IllegalState, how could we get such exception here? (maybe we should check isEmpty earlier in the code)

* https://github.com/jenkinsci/github-branch-source-plugin/blob/f1859729d4e52ad351701dc36bb14ea6cc6b15b5/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java#L145
   * `// may be a user... ok to ignore` are we sure that's it's ok to ignore that we fail to connect to the remote GitHub Server?

* https://github.com/jenkinsci/github-branch-source-plugin/blob/f1859729d4e52ad351701dc36bb14ea6cc6b15b5/src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java#L162
   * `// Something wrong happened, maybe java.net.ConnectException? ` IMHO a network or http exception should not be swallowed applying another behavior, expecially if this alternate behavior is not consistent with the behavior of the `RateLimitExceededException `

* https://github.com/kohsuke/github-api/blob/dbc79f8c423749311495cb548c1a9cda1d8638cf/src/main/java/org/kohsuke/github/GitHub.java#L460
   * an invalid SSL cert or proxy issue will cause an "invalid credentials" message :-(

**Comment by cyrille-leclerc** [comment]
@recena thanks for merging! I shared my opinion too :-)

**Comment by recena** [comment]
@cyrille-leclerc I agree with the motivation of your PR, I only recommend to avoid a lot of ternary operators and magic strings like `github.com`.

If you want we can discuss all these cases when you file the corresponding bugs.

**Comment by jglick** [comment]
Generally any code which looks like

```java
} catch (IOException e) {
    // ignore
}
```

is wrong. You should be logging (perhaps at `FINE`), or catching a more specific type like `FileNotFoundException` if you expect the API to be producing a 404 under certain normal conditions.

**Comment by jglick** [comment]
> I would prefer to have the entire API URL including port, URI & protocol rather than just the hostname.

Agreed.

**Comment by recena** [comment]
@jglick 

> Generally any code which looks like

*Generally* is a way to say *in the majority cases* but when you use GitHub API you will find exceptions.

**Comment by ndeloof** [comment]
@recena spent few months at support and then you"ll change your mind about silently ignoring API errors.

**Comment by recena** [comment]
@ndeloof Thanks for your advice I never thought about it.

**Comment by orrc** [comment]
I'm 100% with you @cyrille-leclerc — I tried using this plugin and got zero feedback about why my config wasn't working, or why my GH Enterprise URL was "wrong".

Thanks for the improvements!

**Comment by recena** [comment]
@orrc Did you file some issue? When did you set the GitHub Enterprise instance the form validation was supported? It was included in `1.1`.

**Comment by recena** [comment]
@orrc I saw you comment [here](https://issues.jenkins-ci.org/browse/JENKINS-33228?focusedCommentId=249824&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-249824). Now, evertything is clear :dizzy_face:

**Pull request merged**